packages/regenerator-transform/src/visit.js

```diff
@@ -177,32 +184,118 @@ function getOuterFnExpr(funPath) {
     node.id = funPath.scope.parent.generateUidIdentifier("callee");
   }
 
+  let outerFnExpr = node.id
+
   if (node.generator && // Non-generator functions don't need to be marked.
-      t.isFunctionDeclaration(node)) {
-    let pp = funPath.findParent(function (path) {
+    t.isFunctionDeclaration(node)) {
+    let pp = funPath.findParent(function(path) {
       return path.isProgram() || path.isBlockStatement();
     });
 
     if (!pp) {
-      return node.id;
+      return funPath;
     }
+    const parameters = funPath.node.params.map(function(param) {
+      return t.cloneDeep(param);
+    })
+
+    // turn our FunctionDeclaration into a FunctionExpression with a different
+    // name - our our eventual outerFnExpr needs to be identical to the original
+    // FunctionDeclaration name so that external calls work correctly, and we
+    // can't also use that name in our FunctionExpression because it would shadow
+    // the outerFnExpr in our wrapping scope
+    const functionExpression = t.functionExpression(
+      funPath.scope.generateUidIdentifierBasedOnNode(node.id),
+      parameters,
+      t.cloneDeep(funPath.node.body),
+      funPath.node.generator,
+      funPath.node.async
+    );
 
-    let markDecl = getRuntimeMarkDecl(pp);
-    let markedArray = markDecl.declarations[0].id;
-    let funDeclIdArray = markDecl.declarations[0].init.callee.object;
-    t.assertArrayExpression(funDeclIdArray);
+    // in spite of the above, we want the resulting function's name property
+    // property to be the same as the original so we wrap the FunctionExpression
+    // with:
+    //   Object.defineProperty(functionExpression, 'name', { value: outerFnExpr.name }
+    // this way introspection / prototype chains work the way we'd expect
```

Comment:
The `regeneratorRuntime.mark` function just returns the function you pass to it, so the two `abc` bindings should refer to the same function object, so the shadowing shouldn't be a problem. Even if the outer value of `abc` gets reassigned, I don't think it's important for `regeneratorRuntime.wrap` to get the new value.

The problem isn't `Object.defineProperty` support (though older IEs mess that up, too), but setting the `name` property on a function object, which either silently fails or throws an exception in some older browsers. It's also just a shame to do something at runtime that could be done at compile time.